### PR TITLE
Adopt rust-prover-tools for formal tooling (15.1.3)

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch implements roadmap task (15.1.3) by adding pinned Kani, Verus, and `rust-prover-tools` metadata plus concise Makefile entry points for local formal-verification tooling. It replaces the earlier plan-only scope with the approved implementation in [docs/execplans/15-1-3-kani-and-verus-metadata-plus-scripts.md](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/docs/execplans/15-1-3-kani-and-verus-metadata-plus-scripts.md).

Roadmap task: (15.1.3)
ExecPlan: [docs/execplans/15-1-3-kani-and-verus-metadata-plus-scripts.md](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/docs/execplans/15-1-3-kani-and-verus-metadata-plus-scripts.md)

It also preserves the requested documentation-maintenance updates by importing the updated [AGENTS.md](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/AGENTS.md) and [docs/documentation-style-guide.md](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/docs/documentation-style-guide.md).

## Review walkthrough

- Start with [Makefile](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/Makefile) and [tools/rust-prover-tools/REF](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/tools/rust-prover-tools/REF) to review the pinned `prover-tools` invocation model.
- Review [tools/kani/VERSION](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/tools/kani/VERSION), [tools/verus/VERSION](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/tools/verus/VERSION), [tools/verus/SHA256SUMS](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/tools/verus/SHA256SUMS), and [tools/verus/README.md](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/tools/verus/README.md) for the tool pins and checksum provenance.
- Review [tests/formal_tooling.rs](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/tests/formal_tooling.rs), [tests/common/formal_tooling_support.rs](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/tests/common/formal_tooling_support.rs), and [tests/features/formal_tooling.feature](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/tests/features/formal_tooling.feature) for the network-free unit and BDD coverage.
- Finish with [docs/developers-guide.md](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/docs/developers-guide.md), [docs/formal-verification-methods-in-wireframe.md](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/docs/formal-verification-methods-in-wireframe.md), and [docs/roadmap.md](https://github.com/leynos/wireframe/blob/338aa26fd1fab7353c884d3d8486c153b32b57ec/docs/roadmap.md) for the documented convention and completed roadmap item.

## Validation

- `make install-kani`: passed; installed and verified `cargo-kani 0.67.0` through `prover-tools`.
- `make check-kani-version`: passed; installed Kani matched `tools/kani/VERSION`.
- `make install-verus`: passed; installed Verus `0.2026.05.24.ecee80a` through `prover-tools` after checksum validation.
- `make run-verus`: expected failure; reported `Verus proof file not found: verus/wireframe_proofs.rs` until later roadmap work adds proofs.
- `make check-fmt`: passed.
- `make lint`: passed.
- `make test`: passed.
- `make markdownlint`: passed.
- `make nixie`: passed.
- `coderabbit review --agent`: final review passed with zero findings.

## Notes

`make run-verus` deliberately remains a clear missing-proof-file failure for 15.1.3 because later roadmap items own `verus/wireframe_proofs.rs`, Kani harnesses, Verus proofs, and formal CI jobs.

## References

- Lody session: https://lody.ai/leynos/sessions/33edb5cf-ed18-43de-b3d5-4dc774ec7610

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[tests/common/formal_tooling_support.rs](https://github.com/leynos/wireframe/pull/530/files/338aa26fd1fab7353c884d3d8486c153b32b57ec#diff-910a842c95b1b6a4ea551b25955f3a9b905b8006f9e82c6134a5f2bf011210b0)

Comment on file

~~~rust
//! Shared helpers for formal-tooling metadata tests.
~~~

❌ New issue: [String Heavy Function Arguments](https://codescene.io/projects/68308/delta?repo-id=1125256&review-id=530&biomarker=String+Heavy+Function+Arguments&filename=tests%2Fcommon%2Fformal_tooling_support.rs&method=)
In this module, 100.0% of all arguments to its 16 functions are strings. The threshold for string arguments is 39.0%

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response. Ensure that this is validated against the *current* version of the codegraph.

If further refinement to address this finding would be deleterious, please supply a clear explanatory one to two paragraph markdown message I can paste into the CodeScene web ui's diagnostic suppression function so this diagnostic can be silenced.

[tests/common/formal_tooling_support.rs](https://github.com/leynos/wireframe/pull/530/files/75dca55b4f3d6bab8c584f67a2e9beed62ab53c5#diff-910a842c95b1b6a4ea551b25955f3a9b905b8006f9e82c6134a5f2bf011210b0)

Comment on file

~~~rust
//! Shared helpers for formal-tooling metadata tests.
~~~

❌ New issue: [String Heavy Function Arguments](https://codescene.io/projects/68308/delta?repo-id=1125256&review-id=530&biomarker=String+Heavy+Function+Arguments&filename=tests%2Fcommon%2Fformal_tooling_support.rs&method=)
In this module, 46.2% of all arguments to its 17 functions are strings. The threshold for string arguments is 39.0%

File: docs/formal-verification-methods-in-wireframe.md [hunk 1]
```diff
@@ -904,24 +931,15 @@ kani-smoke:
     - uses: actions/checkout@v5
     - name: Setup Rust
       uses: leynos/shared-actions/.github/actions/setup-rust@<PINNED_SHA>
-    - name: Install Kani
-      run: scripts/install-kani.sh
+- name: Install Kani
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/formal-verification-methods-in-wireframe.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai pause

**Comment by leynos** [comment]
@coderabbitai pause

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 3 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Rust Compiler Lint Integrity | ❌ Error | tests/common/formal\_tooling\_support.rs exports repo\_root() and repo\_dir() as pub(crate) though only used internally, violating module-boundary narrowing requirements. | Change repo\_root() and repo\_dir() from pub(crate) fn to private fn to narrow module boundaries and preserve compiler dead-code detection capability. |
| Testing (Unit And Behavioural) | ⚠️ Warning | BDD tests disguise unit tests wrapping internal helpers rather than exercising functional boundaries; end-to-end tests missing for Make targets' command-line behaviour and prover-tools integration. | Add end-to-end tests invoking the four Make targets to verify prover-tools CLI integration, exit codes, and failure modes. Expand unit tests for edge cases in version/checksum parsing and error paths. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces validation functions with invariants over input ranges but tests only pinned values; proptest coverage of version/SHA256 format validation is missing. | Add proptest-based property tests validating is\_three\_part\_numeric\_version and is\_sha256\_hex across ranges of valid and invalid inputs to substantively test invariants. |
| Architectural Complexity And Maintainability | ⚠️ Warning | Duplicates repo\_root() and repo\_dir() from workspace\_manifest\_support.rs without reuse or extraction, violating the not-invented-here principle for existing in-repo test infrastructure. | Extract shared repo-access utilities to a common test module and reuse in formal\_tooling\_support.rs to eliminate duplication. |

**Comment by leynos** [comment]
@coderabbitai review

File: docs/formal-verification-methods-in-wireframe.md [hunk 2]
```diff
@@ -814,11 +814,38 @@ entry point for `make verus`, following the same broad pattern as Chutoro’s
...
-targets.[^15] The following targets should be added:
+Wireframe’s top-level `Makefile` now separates tool installation from
+future-proof execution. Roadmap item 15.1.3 adds the local tool entry points:
 
 ```make
-.PHONY: test-verification stateright kani kani-full verus formal formal-pr formal-nightly
+.PHONY: install-kani check-kani-version install-verus run-verus
+
+install-kani: ## Install the pinned Kani verifier
+ $(PROVER_TOOLS) kani install --repo-root .
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/formal-verification-methods-in-wireframe.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 error, 3 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Rust Compiler Lint Integrity | ❌ Error | tests/common/formal\_tooling\_support.rs exports repo\_root() and repo\_dir() as pub(crate) though only used internally, violating module-boundary narrowing requirements. | Change repo\_root() and repo\_dir() from pub(crate) fn to private fn to narrow module boundaries and preserve compiler dead-code detection capability. |
| Testing (Unit And Behavioural) | ⚠️ Warning | BDD tests disguise unit tests wrapping internal helpers rather than exercising functional boundaries; end-to-end tests missing for Make targets' command-line behaviour and prover-tools integration. | Add end-to-end tests invoking the four Make targets to verify prover-tools CLI integration, exit codes, and failure modes. Expand unit tests for edge cases in version/checksum parsing and error paths. |
| Testing (Property / Proof) | ⚠️ Warning | PR introduces validation functions with invariants over input ranges but tests only pinned values; proptest coverage of version/SHA256 format validation is missing. | Add proptest-based property tests validating is\_three\_part\_numeric\_version and is\_sha256\_hex across ranges of valid and invalid inputs to substantively test invariants. |
| Architectural Complexity And Maintainability | ⚠️ Warning | Duplicates repo\_root() and repo\_dir() from workspace\_manifest\_support.rs without reuse or extraction, violating the not-invented-here principle for existing in-repo test infrastructure. | Extract shared repo-access utilities to a common test module and reuse in formal\_tooling\_support.rs to eliminate duplication. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai review
